### ISSUE TITLE
Introduce a coding standard tool

### ISSUE THREAD

**Issue by pamil** [Behat/Behat#1306]
It'd be nice to have some automated coding standard tool to help us maintain a consistent codebase and get rid of ancient `array()` syntax. What coding standard should Behat follow? I'm open to suggestions.

**Comment by gmponos** [comment]
IMHO you should use both codesniffer and PHP-CS-Fixer for different reasons.

**Codesniffer**
- More configurable than php-cs-fixer
- It's not only about "style" but more into introducing a coding standard as you say on the title of your issue. For instance you can ban functions using codesniffer.. like `die` but it will not autofix them..
- The maintainer of the package chooses not to auto-fix rules that he considers risky (at least that the feeling I get).. for instance if there is a rule that you are required to add declare strict types he will not introduce a fix regarding that. But there is slevomat/coding standard that does a lot of them

**PHP-CS-fixer**

- is all about fixing.. rules are related with fixing stuff. I don't think it reports things that can not be fixed.
- not so handly config (or at least that's what I know) if I want to disable a specific rule per file.. or enable a specific rule per file.

nice coding standards:
- https://github.com/doctrine/coding-standard
- https://github.com/slevomat/coding-standard

and also this for php-cs-fixer
https://mlocati.github.io/php-cs-fixer-configurator/#version:2.16

**Comment by Taluu** [comment]
Meh, not sure CS-Fixer is needed. You can always run `CBF` that is installed with Code-Sniffer.

**Comment by gmponos** [comment]
Well.. in a way I agree with @Taluu  .. many things will be covered by codesniffer if you install slevomat

**Comment by Yozhef** [comment]
On my projects I use this analyzer [PHP Code Sniffer](https://packagist.org/packages/squizlabs/php_codesniffer) - it helps to cover many cases + there is an automatic fix of errors.
Ready to help with integration and take on library style improvements.

**Comment by pamil** [comment]
In Sylius, we use ECS with https://github.com/SyliusLabs/CodingStandard, which has only the autofixable coding standard rules included. Most of the issues that should be handled manually are discovered by PHPStan and Psalm.

**Comment by ciaranmcnulty** [comment]
I have a worry around keeping version history understandable - how can we apply a tool sensibly around areas that are changing?

**Comment by rpkamp** [comment]
I would say start with PSR-12 and work up from there if the need arises (maybe Doctrine CS)?

**Comment by bitwombat** [comment]
> I have a worry around keeping version history understandable - how can we apply a tool sensibly around areas that are changing?

Not sure I understand when commits are atomic. Is the worry that merging PRs would be difficult if a analyzer/fix had been run over the code?

**Comment by ciaranmcnulty** [comment]
@bitwombat just that `git blame` gets confusing

**Comment by andrei-petre** [comment]
> @bitwombat just that `git blame` gets confusing

Hi! I found that this problem can be addressed with the git --ignore-revs-file (>= git 2.23)

Some short blog entries with more details:
[https://akrabat.com/ignoring-revisions-with-git-blame/](https://akrabat.com/ignoring-revisions-with-git-blame/)
[https://moxio.com/blog/ignoring-bulk-change-commits-with-git-blame/](https://akrabat.com/ignoring-revisions-with-git-blame/)

**Comment by AlexSkrypnyk** [comment]
@acoulton @carlos-granados 
I suggest adding the PHPCS (with a standard decided here) to CI with `||true` to have it running in one PR. Then add a second PR with all the fixes.
If you could please confirm this process, I can work on both PRs.

**Comment by acoulton** [comment]
@AlexSkrypnyk it'd be great if you want to work on this, but please could you wait a bit? There are still some of the open PRs quite close to merging and I'd rather not create conflicts with them all, so better to get caught up with the backlog first before we start reformatting things. 

There was also a PR #1433 I closed last week that had started work on this, which might be useful to look at / steal from.

And it might be worth looking at https://github.com/easy-coding-standard/easy-coding-standard (if you're not aware of that) to apply some fixes automatically to save manual work.

**Comment by AlexSkrypnyk** [comment]
@acoulton 
Let me ask in another way:
1. What coding standard would be used? Can a decision be made here?
2. How would the maintainers like to see this work done: single PR with tooling and fixes or 2 PRs?
3. What are the preferred tools?

I can work on things once I have these inputs.

Thanks

**Comment by carlos-granados** [comment]
I second the option of using Easy Coding Standard, a really nice tool that makes applying a coding standard a breeze. Created by the same guy that brought us Rector.

Regarding the coding standard I would at least aim for PSR-12, that one seems to be the best known currently. If we are moving towards a PHP8 only version of the code we might consider going to PER-CS 2.0

I am not sure of the value of splitting this work in 2 PRs, so I'll probably go for one

We also discussed that we would need to add a `.git-blame-ignore-revs` file and include this kind of commits so that git blame does not get affected by this kind of refactoring

**Comment by acoulton** [comment]
Happy with PER-CS 2.0.

For process I'd like to see a single PR with:

* A commit that introduces the tooling and rules 
* One or more commits fixing the code (see below)
* Only when the PR is reviewed and ready to merge, a blame-ignore file


If possible I'd like rules applied in stages (unless the diff is very small). Anything that's only to do with whitespace can be all in one commit. Anything more significant (changing array syntax, heredocs, any typehints etc) would ideally be separate commits for each rule/group of similar rules. This will make it easier to review for unexpected changes.

**Comment by stof** [comment]
My own vote would be to use php-cs-fixer as tool (using its `check` command in the CI).
My reasoning (inherited from Fabien's reasoning for Symfony, to be fair) is that if a coding standard rule cannot be automatically fixed, enforcing a rule for it will annoy contributors having to respect it, while the readability benefit might not be huge.
php-code-sniffer supports auto-fixing for some of their rules, but far from all of them. On the other hand, php-cs-fixer cannot add non-fixable rules by design (as what gets implemented is a fixer)

**Comment by AlexSkrypnyk** [comment]
@carlos-granados 
I was waiting for a final decision on the tooling and the process.

I can implement the tooling as per @acoulton comment.
The process, though, will require more time that I do not have: I cannot do separate separate commits per change. I can do one PR with both tooling and code fixes. Then it can be taken apart to be separate commits, if needed. Would that work?

**Comment by carlos-granados** [comment]
@AlexSkrypnyk if you can create a PR with the tools then I can work on adding the changes in separate committs. It's easier to apply the rules one by one or in groups and then create committs than to separate all the changes in individual committs

**Comment by AlexSkrypnyk** [comment]
Sorry guys, I was completely swamped with non-OSS work. I can get to this only in 2-3 weeks. Maybe someone else should pick this up. Apologies.

**Comment by carlos-granados** [comment]
I have added an initial implementation in this PR https://github.com/Behat/Behat/pull/1592
